### ISSUE TITLE
Timestamp sync seems broken

### ISSUE THREAD

**Issue by andresdrb** [JKorf/Binance.Net#1014]
**Describe the bug**
Since upgraded to 8.x version the timestamp sync ssems to intermitently (every few hours) fail and spend 1-2 hours out of sync (givin "Timestamp for this request was 1000ms ahead..." messages).

**To Reproduce**
Happens on all endpoints called (I call accountInfo, exchangeInfo, trading, etc). AutoTimestamp setting is set to true

**Comment by JKorf** [comment]
Hi, are you using the latest version? There was an earlier version where the time offset would not get reset after the difference between the server and the client got smaller than 500ms after being more than 500ms before.

**Comment by andresdrb** [comment]
Yes, using 8.0.3. I think I may found the issue. I noticed that the PartialRateLimiter was kicking off a lot, so I changed to total rate limiter and the issue seems gone. I have the feeling that because the execution of the request was being delayed, those requests didn't get the AutoTimestamp (maybe they were applied the autotimestamp before de Rate Limit wait), hence when they are finally executed they are "out of sync" of timestamp.  Makes sense?

**Comment by andresdrb** [comment]
I have a feeling that my assumption is 100% right. My ratelimiter kick in briefly and right there I got a couple of Timestamp errors again

**Comment by JKorf** [comment]
Right, that makes sense. I'll look into a fix.

**Comment by andresdrb** [comment]
My pleasure 😊

Enviado desde mi iPhone

**Comment by andresdrb** [comment]
Also it feels like applying by default the PartialEndpointRate limiter is too aggressive, specially if you use a mix of endpoints. In my vase it was triggering all the time on the get account info endpoint, which is an expensive one, but making way more calls on lighter endpoints. It feels like the totalRateLimiter is a better choice by default.

Enviado desde mi iPhone

**Comment by andresdrb** [comment]
Notice how the performance of some of my code degraded dramatically because of the rate limiter since I upgraded to 8.0.3 on friday. It went back to reasonable times today when I changed to the totalRateLimiter, with no errors or failures
![image](https://user-images.githubusercontent.com/9131538/155895084-59640a9c-ea2f-4791-a35b-eeaa1973f46b.png)

**Comment by JKorf** [comment]
Hm intresting, I'll take a look at that. Also I've fixed the issue where timestamps time out when the rate limiter triggers in 8.0.4. Although that might not be an issue for you anymore if it no longer triggers ;)

**Comment by andresdrb** [comment]
Yeah for me it is fine as I changed to TotalRateLimiter, but yeah definitelly will help others (It was a bit tricky to find the root cause and there may be people struggling with it out there). Thanks!

**Comment by andresdrb** [comment]
Hey, how did you test the fix? I got a burst of timestam errors for 1 hour earlier today after upgrading to 8.0.4...

**Comment by JKorf** [comment]
Hm weird. I've tested it by adding an `await Task.Delay(10000)` in the spot where the rate limit would trigger, so simulating a rate limit waiting. That was correctly creating a timestamp after the rate limiting had passed.
Was it happening again after a rate limit? Or during normal process? If you have a lot of time drift, it might be that time syncing once every hour isn't often enough.

**Comment by andresdrb** [comment]
Ah! So autotimestamp by default syncs every hour? It may be that….

Enviado desde mi iPhone

**Comment by JKorf** [comment]
Yes, you can set the interval in the client options. But if you have a lot of drift the best solution might be to use a third party app to sync your server time more often with a time provider. People have been using the SP TimeSync tool with success to make sure the the time is correct, which will always be a more reliable way than trying to keep in sync by offsetting the timestamp in this library.

**Comment by andresdrb** [comment]
Yeah my problem was I removed the autotimestamp interval adjustment in the options when I was investigating the rate limiter issue 😅 the issue with trying to update the OS time is that I’m running my stuff in azure app service and when I looked at it some time ago it was not very reliable to adjust the os time on those types of virtual machines

Enviado desde mi iPhone

**Comment by JKorf** [comment]
Haha I see. Try using a shorter interval, and let me know if you have any more issues

**Comment by andresdrb** [comment]
I’m going to switch the logging back to debug and let you know as I still some spikes in timestamp errors when i get some spikes in activity. I’ll let you know if I see anything odd

Enviado desde mi iPhone

**Comment by andresdrb** [comment]
I think there's still some cause-effect somewhere between the Timestamp adjustment and those failures I see. See how on the first image the timestamp is supposed to be adjusted at 1.45PM and right after taht I got a bunch of timestamp errors (when the timestamp was supposed to be adjusted just a moment before). Also it is interesting that it happens mainly on my calls to the GetAccountInfo, just in case that gets you some more insight...

![image](https://user-images.githubusercontent.com/9131538/155995228-9dab71d6-8034-4f8c-a07b-2e75e9d1f4ee.png)

![image](https://user-images.githubusercontent.com/9131538/155995394-dd75fc12-f5d0-40f5-b163-8b64c7345bac.png)

**Comment by JKorf** [comment]
Hm well the time syncing is an estimate. It's a simple procedure of recording the local timestamp, requesting the server timestamp and then comparing the 2, and adding the difference to the timestamp sent to the server for authenticated requests. Here's the source for determining the offset:
https://github.com/JKorf/CryptoExchange.Net/blob/4c4cfbb60e0def028c074bb6a7459f019fb1c0ae/CryptoExchange.Net/Clients/RestApiClient.cs#L63

The reason you get the errro mainly on the GetAccountInfoAsync call is because timestamping is only used in authenticated requests.

You're running this on Azure, is timestamping necessary? I would excpect Azure server to have a correct timestamp.

**Comment by andresdrb** [comment]
Yeah using it I. Azure but somehow there are misalignments on the timestamps. I believe th GetAccountInfo are authenticated calls aren’t they (you are retrieving one account data, so it has to be authenticated for that account). Maybe that’s the problem? Also I don’t think the problem is with your time stamp adjustment routine, either those calls to GetAccountInfo need to eb adjusted timestamp too (I think that’s the key, and also I thought I had the feeling you were not adjusting them when I looked at your fix’s changes) or has something to do with volume/concurrency (I see it happen mainly when there are many calls to that method happening at the same time for different users)

Enviado desde mi iPhone

**Comment by andresdrb** [comment]
I found again another occurrence of that strong correlation: whenever I see a timestamp adjustment in the logs because if some other method call, right after that the GetAccountInfo calls fail for a little while. When there has not been any recent timestamp adjustment they are fine



Enviado desde mi iPhone

**Comment by JKorf** [comment]
Those differences seem way too large to me. 1100ms difference between 6 minutes, thats not right. The question is what is the cause of this.. 

I've thought of another potential cause of delayed requests. 
`ServicePointManager.DefaultConnectionLimit`
This property determines the max amount of concurrent connections the application can have. It's default 2 or 10 depending on the type of application. If you're doing concurrent requests it might delay requests and mess with the timestamping/time syncing. 

Can you try upping this value if you haven't already?

**Comment by andresdrb** [comment]
Default connection limit was indeed a problem long ago on my code, but it's been months since that was tuned. This behaviour only started happening since the upgrade to 8.X. Has the way in which HTTP clients are handled changed?

**Comment by andresdrb** [comment]
Or any concurrency race condition that you can think of?

**Comment by JKorf** [comment]
Hm request creating/sending hasnt changed much, rate limiting changed but shouldn't influence it much. Though you could try to turn the rate limiter of completely. 
Timesyncing also changed a bit, though it's mostly just the code moved from Binance to the CryptoExchange.Net base library. Obviously since there seems to be something going on with time syncing it's possible something did change and is breaking now. 
What exact version were you on where you didnt have issues? I'll look into it some more tomorrow.

**Comment by andresdrb** [comment]
I'm going to check on my code for any potential things that may be clogging up connections, but this has not been a problem for long (maybe something has been developing for some time and by coincidence started manifesting itself now).. Can you think of anything that maybe has made the clients more chatty?

**Comment by andresdrb** [comment]
I was on 7.x 😅 I only found out when exchange info started breaking last week. I’ve done some optimisations in my code as it seems that over time I was calling way too much to GetAccountInfo (which is an expensive 10 points weight call anyway), and it feels it could be contributing to it. I’ll let you know if I still see the issue.

Thanks!

Enviado desde mi iPhone

**Comment by JKorf** [comment]
I did find another thing, in 7.x there was some functionality that would re-sync the timestamp when a timestamp error was received. This might prevent multiple errors happening for a certain time span since the time should be re-synced when the error is received. I've re-added this functionality in 8.0.5, so that might help you as well.

**Comment by andresdrb** [comment]
Mmmmm that may be part of it too. After my optimisations since yesterday In my code the occurrences have been reducen massively (tipical thing that you start overloading little by little over time and only manifests itself during the perfect storm hahahaa), but I think your changes in 0.4 and 0.5 will help too 😊 the last one about resynching after a failure makes a lot of sense. Thanks a lot for all this man!

Enviado desde mi iPhone

**Comment by JKorf** [comment]
I know what you mean haha.
All good, let me know if you're having any issues I can help with

**Comment by HypsyNZ** [comment]
Manually change the local time to something absurd then run `w32tm /resync` as administrator

**Comment by HypsyNZ** [comment]
```cs
            var syncLocalTime = new ProcessStartInfo();
            string syncTimeW32 = "w32tm /resync";
            syncLocalTime.UseShellExecute = true;
            syncLocalTime.WorkingDirectory = @"C:\Windows\System32";
            syncLocalTime.FileName = @"C:\Windows\System32\cmd.exe";
            syncLocalTime.Verb = "runas";
            syncLocalTime.Arguments = "/c " + syncTimeW32;
            syncLocalTime.WindowStyle = ProcessWindowStyle.Hidden;
            Process.Start(syncLocalTime);
```

I'm going to start doing this when the time drifts too much

**Comment by JKorf** [comment]
Interesting, this triggers the windows time syncing? Might be a good alternative to a 3rd party program

**Comment by HypsyNZ** [comment]
> Interesting, this triggers the windows time syncing? Might be a good alternative to a 3rd party program

Yes this triggers the built in time-syncing mechanism of the OS, Its basically like clicking "Sync Now" but there is no cooldown.

> Are you sure that will work in PaaS services?

https://docs.microsoft.com/en-us/azure/virtual-machines/windows/time-sync

It seems like you have access to `w32tm`

**Comment by andresdrb** [comment]
As mentioned earlier, is you try that at most you would very temporarily syncing the VM clock, but azure would resync again with the host clock by design, so you would have to constantly update while getting all kind of errors on timestamp with Binance because of the constant fluctuations in clock adjustment. You’d better stick to JKorf‘s timestamp resync 😉

Enviado desde mi iPhone

**Comment by HypsyNZ** [comment]
```
Azure hosts are synchronized to internal Microsoft time servers that take their time from Microsoft-owned Stratum 1 devices, with GPS antennas
```

Yeah if you are using Azure you should just stick to the default implementation of the timestamp sync, it seems like you already have the best UTC timestamp you are going to get

just adding input for others who may read this having similar issues.

**Comment by HypsyNZ** [comment]
https://docs.microsoft.com/en-us/azure/virtual-machines/windows/time-sync

>Virtual machine interactions with the host can also affect the clock. During [memory preserving maintenance](https://docs.microsoft.com/en-us/azure/virtual-machines/maintenance-and-updates#maintenance-that-doesnt-require-a-reboot), VMs are paused for up to 30 seconds. For example, before maintenance begins the VM clock shows 10:00:00 AM and lasts 28 seconds. After the VM resumes, the clock on the VM would still show 10:00:00 AM, which would be 28 seconds off. To correct for this, the VMICTimeSync service monitors what is happening on the host and prompts for changes to happen on the VMs to compensate.

>The VMICTimeSync service operates in either sample or sync mode and will only influence the clock forward. In sample mode, which requires W32time to be running, the VMICTimeSync service polls the host every 5 seconds and provides time samples to W32time. Approximately every 30 seconds, the W32time service takes the latest time sample and uses it to influence the guest's clock. Sync mode activates if a guest has been resumed or if a guest's clock drifts more than 5 seconds behind the host's clock. In cases where the W32time service is properly running, the latter case should never happen.

**Comment by HypsyNZ** [comment]
have you tried turning on `VMICTimeSync service`

**Comment by HypsyNZ** [comment]
have you tried turning on `VMICTimeSync service` or in the case that it is already on, turning it off?

**Comment by HypsyNZ** [comment]
https://github.com/HypsyNZ/TimeCompare

There was a more dramatic difference between the tests the first time I ran this but I didn't copy the result

You should be able to test difference with this even though I have modified the TimeSync stuff

```
Server: 63781797602432| Local: 63781797602426| TestOffset: (6) CalculatedTimeOffset: 5.2038
Server: 63781797602912| Local: 63781797602826| TestOffset: (86) CalculatedTimeOffset: 85.0595
Server: 63781797603186| Local: 63781797603101| TestOffset: (85) CalculatedTimeOffset: 85.53
Server: 63781797603741| Local: 63781797603656| TestOffset: (85) CalculatedTimeOffset: 84.9378
Server: 63781797604827| Local: 63781797604742| TestOffset: (85) CalculatedTimeOffset: 84.7455
Server: 63781797606453| Local: 63781797606367| TestOffset: (86) CalculatedTimeOffset: 85.7431
Server: 63781797608490| Local: 63781797608402| TestOffset: (88) CalculatedTimeOffset: 87.0235
Syncing Local Time
W32API: Synchronized Local Time..
Server: 63781797609976| Local: 63781797609891| TestOffset: (85) CalculatedTimeOffset: 84.6152
Server: 63781797611079| Local: 63781797610992| TestOffset: (87) CalculatedTimeOffset: 86.1629
Server: 63781797611626| Local: 63781797611539| TestOffset: (87) CalculatedTimeOffset: 86.2373
Server: 63781797611900| Local: 63781797611815| TestOffset: (85) CalculatedTimeOffset: 85.6251
Server: 63781797612176| Local: 63781797612090| TestOffset: (86) CalculatedTimeOffset: 85.7572
Server: 63781797612325| Local: 63781797612238| TestOffset: (87) CalculatedTimeOffset: 86.0426
Server: 63781797612475| Local: 63781797612387| TestOffset: (88) CalculatedTimeOffset: 87.1977
Server: 63781797612625| Local: 63781797612539| TestOffset: (86) CalculatedTimeOffset: 85.9239
Server: 63781797612775| Local: 63781797612688| TestOffset: (87) CalculatedTimeOffset: 86.6198
Server: 63781797612923| Local: 63781797612837| TestOffset: (86) CalculatedTimeOffset: 85.0415
Server: 63781797613076| Local: 63781797612990| TestOffset: (86) CalculatedTimeOffset: 85.0575
Server: 63781797613347| Local: 63781797613262| TestOffset: (85) CalculatedTimeOffset: 84.3535
Server: 63781797613623| Local: 63781797613536| TestOffset: (87) CalculatedTimeOffset: 86.465
```

You'll notice from this that syncing your time when Its already within your ping time is pointless

You'll also notice from this that the first request is always faster

**Comment by andresdrb** [comment]
@JKorf  I think I have found somethig that has changed since 8.x that is surely impacting this. 

I kind of saw the other day in the logs that you are automatically "updating/keeping" trading rules" (so making regular exchangeInfo requests to Binance API)  which were not there before. I can see that in the logs of my azure services too, where there are calls being made to the ExchangeInfo endpoint that were not there before. 8.x. IMO you shoul not be automaticaly making those calls (or at least not that frequently) as those are quite expensive and slow and thus more than likely being the ones causing the clogging in outbound requests and so increasing the likelyhood of timestamp missalignment (I know from experience as I was doing a similar thing in the past untill I started caching those on my side with an expiry date of 24 hours).

Notice that I moved to 8.x version by the 25th and that moment is when the exchangeInfo requests spiked

![image](https://user-images.githubusercontent.com/9131538/156471370-60397dd7-dd4e-4538-acc3-8c6bcad8e9ff.png)


Does this make sense?

**Comment by andresdrb** [comment]
It seems that for every order placed you are checking the trade rules, where you are triggering the ExchangeInfor call, suuuuuper expensive and slow call. You shouldn't be doing that unless you are applying some caching

![image](https://user-images.githubusercontent.com/9131538/156471838-d2fc1cb9-f702-4763-9aab-3387b8df7ca1.png)

**Comment by andresdrb** [comment]
ok, last one for today today. So I see you have this option to define how often to refresh the trade rules (which you have set by default to 60 minutes, which is way too often IMO) , so I have set it to 24 hours in my code. This all makes sense now why I was seeing those spikes in failures when there was an spike in orders created from my code. This behaviour should be documented somewhere so people is aware of it.

![image](https://user-images.githubusercontent.com/9131538/156472492-fd6073c0-2625-4048-8e3f-a1db05a06499.png)


![image](https://user-images.githubusercontent.com/9131538/156472280-49f2d5fd-382d-4cf1-becc-8c0928ac5411.png)

**Comment by HypsyNZ** [comment]
personally I think the "Trade Rules" should be the responsibility of the application not the library but that is just my 2 cents.

**Comment by JKorf** [comment]
@HypsyNZ I agree, which is why it isn't enabled by default. Which raises the question, are you setting the `TradeRulesBehaviour` property on the client options to anything other than `None`? Since if it is None (default) it shouldn't be requesting trade rules at all.

**Comment by JKorf** [comment]
@andresdrb Just checking in, how is it atm? Any improvement?

**Comment by andresdrb** [comment]
It happens less, but still happens every now and then. I still feel there's something off that was not happening before...

**Comment by devbar** [comment]
Hi guys, I have the same issue. I am on 8.0.9 and everything runs fine on my dev machine. But when I try to deploy I run into the issue instantly.

I tested this on two different machines. I synced time several times, no way out :(

**Comment by HypsyNZ** [comment]
> But when I try to deploy I run into 

Deploy to where

can you run this and paste to outputs

https://github.com/HypsyNZ/TimeCompare

**Comment by danielhw2** [comment]
I'm also getting this error. In the creation and connection sequence everything works fine, but sometimes, after 15 minutes or so, any operation returns the error "**Timestamp for this request was 1000ms ahead of the server's time.**"... 

The command line:
`ClienteRest.CoinFuturesApi.Account.GetPositionInformationAsync(null, null, RECEIVE_WINDOW_VALUE,  CancellationToken.None).Result;`

I'm using the latest versions:
Binance.net 8.0.11
CryptoExchange.net 5.1.7

**Comment by JKorf** [comment]
Cleaning up old issues, if you feel this issue is still relevant feel free to comment.
